### PR TITLE
DAOS-15670 vos: SV overwrite missed tx_add_range()

### PR DESCRIPTION
**Pull request by NiuYawei**
In SV overwerite case, the btr_update_record() will defer free the original record and allocate new record for record replacing, however, btr_node_tx_add() is mistakenly skipped in btr_update(), that leads to:
1. In md-on-ssd mode, tree node changes are missed in WAL.
2. In pmem mode, tree node snapshot is missed in undo log.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/vos/vos_tree.c [hunk 1]
```diff
@@ -690,7 +690,12 @@ svt_rec_update(struct btr_instance *tins, struct btr_record *rec,
 	if (rc != 0)
 		return rc;
 
-	return svt_rec_alloc_common(tins, rec, skey, rbund);
+	rc = svt_rec_alloc_common(tins, rec, skey, rbund);
+	if (rc != 0)
+		return rc;
+
+	/* Inform btr_update() that the original record is replaced */
+	return 1;
```
Comment (liw) [inline src/vos/vos_tree.c:698]: [Stupid question] So `svt_rec_update` will never return 0? Can't `btr_update` just handle `rc == 0` case (suppose that we don't make any change to `svt_rec_update`)?

File: src/common/btree.c [hunk 2]
```diff
@@ -2279,6 +2279,10 @@ btr_update(struct btr_context *tcx, d_iov_t *key, d_iov_t *val, d_iov_t *val_out
 				goto out;
 		}
 
+		/* Original record is replaced by above btr_rec_update() */
+		if (rc == 1)
```
Comment (liw) [inline src/common/btree.c:2283]: The `btr_node_tx_add` call may have assigned 0 to `rc`.

Comment (NiuYawei) [reply src/common/btree.c:2283]: oops, good catch.

File: src/vos/vos_tree.c (hunk 1 above)
Comment (NiuYawei) [reply src/vos/vos_tree.c:698]: For svt_rec_update() (registered for SV tree tyoe), it only returns -DER_NO_PERM or 1, but other tree type may register ".to_rec_update()" returning 0. (for in-place overwrite or just skip the update operation).

Comment (liw) [reply src/vos/vos_tree.c:698]: Ah, indeed. I would consider updating the API doc of `to_rec_update` to highlight this new return value.

Comment (NiuYawei) [reply src/vos/vos_tree.c:698]: missed this reply before pushing, I'll update the comment if the patch needs be refreshed (it's highly likely). thanks!

**Review by liw — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**